### DISCUSSION TITLE
Should my Models have Observable Properties?

### DISCUSSION THREAD

**Discussion by Peaj** [dotnet/maui#16118] [Q&A]
# Problem

I am just starting out with MAUI and was using the MVVM Community Toolkit as recommended in various tutorials.
I was pretty happy with my Model logic being completely clean and not using any MAUI specific code at all.

But than I encountered the problem that observing a complete object does not work as I expected.
```CSharp
[ObservableProperty]
private WorkTimeTracker workTimeTracker = new();
```
Using this code and Binding to `WorkTimeTracker.StartTime` does not seem to update the UI parts that are displaying `WorkTimeTracker .WorkTime` etc.

# Solution?

I can only see two solutions to this problem:

## 1) Using observable property attributes in the Model
  - Dirties the Model code with MAUI specific attributes
  - Binding to `WorkTimeTracker .StartTime` also raises the question whether the View should know about the Model at all

## 2) Wrap every Model property in a ViewModel property and call `OnPropertyChanged`

```CSharp
public partial class EditWorkTimesViewModel : BaseViewModel
{
  ...
  public TimeSpan WorkTime
  {
      get => this.WorkTimeTracker.WorkTime;
      set
      {
          this.WorkTimeTracker.WorkTime = value;
          OnPropertyChanged(nameof(WorkTimeTracker));
          OnPropertyChanged(nameof(StartTime));
      }
  }
  ...
}
```

- Seems to get pretty tedious
- Has the problem that you need to know the implementation details of `WorkTimeTracker` in the ViewModel to know what changed

Maybe I am just missing something here. But most of the Tutorials seem to just stop before the Model gets too complex.
Or maybe I am just being too puristic here?

**Comment by ne0rrmatrix** [comment]
If you are trying to update an object inside a ObservableCollection after page has loaded you can try something like the following.
Here is an example using community toolkit for Maui.

```csharp

namespace NerdNewsNavigator2.Model;

/// <summary>
/// A class for storing <see cref="Download"/> in a Database.
/// </summary>
[Table("Downloads")]
public partial class Download : ObservableObject
{
   [PrimaryKey, AutoIncrement, Column("Id")]
    public int Id { get; set; }
    [ObservableProperty]
    private string _copyRight;
    [ObservableProperty]
    private string _fileName;
    [ObservableProperty]
    private bool _deleted;
}

```
That above is part of model file that I use. Also adding:

```csharp
 /// <summary>
    /// An <see cref="ObservableCollection{T}"/> of downloaded <see cref="Download"/> managed by this class.
    /// </summary>
    [ObservableProperty]
    private ObservableCollection<Download> _downloadedShows;
```
Adding  ```[ObserrvableProperty]``` lets you dynamically update an observable object or observable collection. Just adding observable will not apply property changed outside of adding or removing an item to it. If you want to update properties of a collection you need to add observable property to both the collection and model.

**Reply by Peaj** [reply]
Thank you for your answer.
I was not asking about ObservableCollections though.

Maybe I was not specific enough about my model.
WorkTimeTracker is my model class and I was asking if I should/have to use [ObservableProperty] inside this model class.
I was under the impression that I could manage all the bindings in the view model which would leave the model nice and clean
with no MAUI specific dependencies.

From your answer I can see that you are using [ObservableProperty] in your models class.
I have also since found some code examples doing the same so it seems to be the way to go.

Although in your case of using an ObservableCollection I am wondering if it wouldnt be more MVVM conform to have a DownloadViewModel for each list item because you are technically skipping the view model when binding to download.FileName etc.
Sometimes there may even be UI specific data to be stored with the list items like "IsSelected" etc.

**Reply by ne0rrmatrix** [reply]
Sorry about the confusion. I am using a collection view to display a list of items that are wrapped in Observable collection. I have to update the properties in the collection itself from an event that is in a background process.

 share some model data across pages. An example is I have a button you can click to download a show from a podcast. That show can appear on the most recent show list, which is on a different page. I have a cancel button that appears once download start. The show also has download button go away and so does cancel button when download completes.

 When you click on download it spawns a background thread that starts the download and uses events to update the relevant button on what ever page I am on for the current item in the observable collection I am using. So updating an item in collection that is being actively updated from a background process is what I need to do. So decorating the model with ```[ObservableProperty]``` was the first solution I came across that worked.

**Reply by Peaj** [reply]
I see. The background thread might make it trickier. But I guess there should still be a way to wire it up via a DownloadViewModel that is shared across your pages. For now I also put [ObservableProperty] attributes in my model to get it to work. But I will see if I can get rid of them in the future.

**Comment by sjordanGSS** [comment]
Of the two options presented, I've often taken the second route. Although tedious it's pretty straightforward and provides an opportunity to pull in data from other sources when creating the ViewModel.

As an aside for something I've seen elsewhere in this thread, you don't need an ```ObservableProperty``` for an ```ObservableCollection<T>```. It's more efficient to do something like 
```
public ObservableCollection<string> Strings { get; } = new();
```
so it only gets bound to once and then use ```Add()``` and ```Clear()``` than it is to keep creating and binding to new ```ObservableCollection```s. You can even use ```Mode=OneTime``` on the ```Binding``` in this case to make that a little bit more efficient as well.

**Reply by Peaj** [reply]
Thanks for the insight.
That would also mean that the list item should be a view model and not the model directly.
Interestingly most MVVM examples just seem to ignore that.

**Comment by SalKly** [comment]
I am just starting out with MAUI and I have the same exact question so is still using [ObservableProperty] inside of  the model is the way to go ? it doesn't feel right D:

**Comment by slabarque** [comment]
The [MVVM ](https://learn.microsoft.com/en-us/dotnet/architecture/maui/mvvm) article provides some insights on this topic:

>  The view model might choose to expose model classes directly to the view so that controls in the view can data bind directly to them. In this case, the model classes will need to be designed to support data binding and change notification events.
Each view model provides data from a model in a form that the view can easily consume. To accomplish this, the view model sometimes performs data conversion. Placing this data conversion in the view model is a good idea because it provides properties that the view can bind to. For example, the view model might combine the values of two properties to make it easier to display by the view.

So you either take the elegance hit and pollute your models with Observable stuff. This means the responsibility of your models is now not only to store data but also to update the view, making them well, viewmodels. Or you convert back and forth between your models and separate models that can be consumed by the view.

**Comment by lloydjatkinson** [comment]
I don't like the idea of polluting anything not UI/MVVM related with MVVM type attributes. Instead, I wrap the "model" (really a class from another assembly) with an `ObservableObject` like this. I also implement the original interface my "model" implements to ensure that nothing gets missed. Wrapping may not work for all use cases, and there is obviously some repetition, but a source generator may be one solution to that. For me, the cost of mild duplication is worth [not making my pure business logic depend on UI concerns](https://blog.ploeh.dk/2023/10/09/whats-a-sandwich/).

I quote model because you can use any word here, service, manager, whatever.

I really like this answer on Stack Overflow: https://stackoverflow.com/questions/42463886/should-i-use-observablecollections-in-the-model-in-mvvm

> It depends on what a model really is in this case, or rather how you define a model.
> If it is some kind of domain business object that is defined in an assembly that is referenced by your business or service layer and may be used across your entire domain, it shouldn't implement any kind of client-specific interfaces such as INotifyCollectionChanged.

> Then you should prefer to use a generic type such as IList and then create a wrapper class in the client application that you bind to the view element to.
> But if the model is a class that is used only inside your client application and not in any other tiers of your application, then you might as well define the collection property as an ObservableCollection and bind to this one directly.

> The point is that a domain business object should not have any knowledge of WPF or the fact that a client application may bind to it. That's why it seldom makes sense to bind directly to such objects in a WPF application without wrapping them in WPF-aware view model classes first.
> This does add some extra work and an additional class but at the same time it helps maintain separation of concerns which is usually good for maintenance.

```csharp
public class ObservableAlgorithm : ObservableObject, IAlgorithm
{
    private readonly IAlgorithm algorithm;
    private Cell? currentCell;

    public Cell? CurrentCell
    {
        get => this.currentCell;
        set => this.SetProperty(ref this.currentCell, value);
    }

    public ObservableAlgorithm(IAlgorithm algorithm)
    {
        this.algorithm = algorithm;
    }

    public void Compute()
    {
        this.algorithm.Compute();
        this.CurrentCell = this.algorithm.CurrentCell;
    }

    public bool ComputeByStep()
    {
        bool result = this.algorithm.ComputeByStep();
        this.CurrentCell = this.algorithm.CurrentCell;
        return result;
    }
}
```
